### PR TITLE
Build4linux

### PR DESCRIPTION
**Pull request by JuergenKosel**
Hello,

some modifications are necessary to build iot-gateway-proxy for Debian 8.7 (Jessie) with the call of the build script:

`./build.sh --config Release --skip-unittests --skip-dotnet`

The major change is needed, because the sub-project libwebsocket can not be compiled with gcc option --std=c99 and warnings threaded as errors.
Also some type and macro definitions need to be changed.

Greetings
  Juergen

### CODE REVIEW

File: bld/CMakeLists.txt [hunk 1]
```diff
@@ -66,11 +66,13 @@ endif()
 
 if (CMAKE_VERSION VERSION_LESS "3.1")
```
Comment (marcschier) [inline bld/CMakeLists.txt]: It would be better to change this to a macro, to avoid repeating the stuff across all CMakeList.txt files.  It used to be that this was a macro, named compileAsC99, that might be good to go back to.  I am surprised though that this does not happen on Ubuntu (with gcc 4.8, 6.2.0.*).  What was the compiler you used on Debian?  Another option, which I may prefer is to unset the c standard before including libwebsockets and re-setting it after (like done with warnings).  Might not work though, cmake is finicky with these variables...

File: bld/build.sh [hunk 2]
```diff
@@ -94,7 +94,8 @@ native_build()
 			cmake -DCMAKE_BUILD_TYPE=$c -Dskip_unittests:BOOL=$skip_unittests \
 					-Duse_zlog:BOOL=$use_zlog "$repo_root" || \
 				return 1 
-			make || \
+                        # Use all available CPU cores for make in silent mode. But be verbose if something goes wrong
+			make -j`nproc` || make VERBOSE=1 || \
```
Comment (marcschier) [inline bld/build.sh]: Nice change, thanks!

File: bld/src/libiotsdk/CMakeLists.txt [hunk 3]
```diff
@@ -6,6 +6,22 @@
...
+	  set (CMAKE_C_FLAGS "--std=c99 ${CMAKE_C_FLAGS}")
+          set (CMAKE_CXX_FLAGS "--std=c++11 ${CMAKE_CXX_FLAGS}")
+          # Macro _POSIX_C_SOURCE is needed by ${_SHARED_ROOT}
+          set (CMAKE_C_FLAGS "-D_POSIX_C_SOURCE=200112L ${CMAKE_C_FLAGS}")
+	endif()
+else()
+	set (CMAKE_C_STANDARD 99)
+    set (CMAKE_CXX_STANDARD 11)
+          # Macro _POSIX_C_SOURCE is needed by ${_SHARED_ROOT}
+          set (CMAKE_C_FLAGS "-D_POSIX_C_SOURCE=200112L ${CMAKE_C_FLAGS}")
```
Comment (marcschier) [inline bld/src/libiotsdk/CMakeLists.txt]: Principally nothing against the POSIX macro (although would like to avoid defining it for non posix builds).  What was the error you encountered?  Maybe I rather fix this in a targeted way in the shared project then..

File: bld/CMakeLists.txt (hunk 1 above)
Comment (JuergenKosel) [inline bld/CMakeLists.txt]: Debian 8.7 provides gcc version 4.9.2 (Debian 4.9.2-10).
But I believe, the statement above checks for version of cmake, which is 3.0.2.
Do we need to check the version of gcc or cmake here?

File: src/pal/pal_sk_posix.c [hunk 4]
```diff
@@ -1195,7 +1195,7 @@ int32_t pal_socket_getsockopt(
     int32_t opt_lvl, opt_name;
     socklen_t opt_len;
     int32_t opt_val;
-    u_long avail;
+    prx_size_t avail;
```
Comment (marcschier) [inline src/pal/pal_sk_posix.c]: Yeah, good one, what I could find is that IOCtl takes a pointer to an int, so prx_size_t should work.  However, I might decide to typedef prx_size_t to size_t or even replace it. so, I would say u_long might better change to int, then assinging it to *value should be fine.

File: src/pal/pal_sk_posix.c [hunk 5]
```diff
@@ -1389,7 +1389,9 @@ void pal_socket_free(
     mem_free_type(pal_socket_t, sock);
 }
 
+#ifndef USE_CTEST // avoid conflicting definition of clockid_t in os_mock.h:882 and /usr/include/time.h
```
Comment (marcschier) [inline src/pal/pal_sk_posix.c]: I would rather like to fix this in os_mock.h then if we absolutely cannot avoid including time.h at all e.g. because it comes through standard headers on the distro.  Then I would also try to find the actual version of the header or some other identifying system info to preprocess the clock_id type out of os_mock.h.  If possible.

File: test/ut/pal/new__pal_sk_posix_ut.c [hunk 6]
```diff
@@ -43,7 +43,7 @@ MOCKABLE_FUNCTION(, int32_t, close,
     fd_t, fd);
 // ioctl
 MOCKABLE_FUNCTION(, int, ioctl,
-    fd_t, fd, unsigned int, cmd, u_long*, arg);
```
Comment (marcschier) [inline test/ut/pal/new__pal_sk_posix_ut.c]: This should be "int"

**Comment by marcschier** [comment]
Hi @JuergenKosel, thank you so much for reporting this and your contributions you provided to help fix the issue.  Please see my review comments.  

It would be great if you could add a "build" DockerFile that reflects the setup of your box and that you used when you encountered the build error.  That way I can easily repro any future breaks or issues, without keeping my own distro installed and up and running.  The build DockerFiles are in bld/Docker/...  See the windows build.cmd for how they are used.  We need something similar for build.sh ;-)

If you do not have time, no worries, please add an issue (in this case listing all packages installed on your box, and compiler version used) and I can take it from there as well.  I can then also bring in the other fixes if you want to.  

Best regards,

Marc

File: bld/src/libiotsdk/CMakeLists.txt (hunk 3 above)
Comment (JuergenKosel) [inline bld/src/libiotsdk/CMakeLists.txt]: It is already in the shared project:
https://github.com/Azure/azure-c-shared-utility/blob/master/CMakeLists.txt#L177
But the CMakeList.txt of the shared project is not used.
So I get the following output without this macro set here:

make[2]: Entering directory '/home/klj/iot-gateway-proxy/build/cmake/Release'
/usr/bin/cmake -E cmake_progress_report /home/klj/iot-gateway-proxy/build/cmake/Release/CMakeFiles 
[  0%] Building C object bld/src/libiotsdk/CMakeFiles/libiotsdk.dir/__/__/__/deps/shared/adapters/condition_pthreads.c.o
cd /home/klj/iot-gateway-proxy/build/cmake/Release/bld/src/libiotsdk && /usr/bin/cc  -DNO_ZLOG -DUSE_OPENSSL=1 --std=c99  -O3 -DNDEBUG -fPIC -I/home/klj/iot-gateway-proxy/deps/shared/inc    -o CMakeFiles/libiotsdk.dir/__/__/__/deps/shared/adapters/condition_pthreads.c.o   -c /home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c:22:1: error: unknown type name ‘clockid_t’
 clockid_t time_basis = -1;
 ^
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c: In function ‘set_time_basis’:
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c:40:5: warning: implicit declaration of function ‘pthread_condattr_getclock’ [-Wimplicit-function-declaration]
     pthread_condattr_getclock(&cattr, &time_basis);
     ^
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c: In function ‘create_cond’:
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c:56:9: warning: implicit declaration of function ‘pthread_condattr_setclock’ [-Wimplicit-function-declaration]
         pthread_condattr_setclock(&cattr, time_basis);
         ^
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c: At top level:
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c:65:24: warning: ‘struct timespec’ declared inside parameter list
 int get_time_ns(struct timespec* ts)
                        ^
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c: In function ‘get_time_ns’:
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c:85:5: warning: implicit declaration of function ‘clock_gettime’ [-Wimplicit-function-declaration]
     err = clock_gettime(time_basis, ts);
     ^
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c: In function ‘Condition_Wait’:
/home/klj/iot-gateway-proxy/deps/shared/adapters/condition_pthreads.c:140:29: error: storage size of ‘tm’ isn’t known
             struct timespec tm;
                             ^
bld/src/libiotsdk/CMakeFiles/libiotsdk.dir/build.make:721: recipe for target 'bld/src/libiotsdk/CMakeFiles/libiotsdk.dir/__/__/__/deps/shared/adapters/condition_pthreads.c.o' failed

Comment (marcschier) [inline bld/src/libiotsdk/CMakeLists.txt]: Super, yes, then I would say, let's do what you suggest but scope it in a way it only applies to posix:

```
if(LINUX) // Need a posix flag here...
     set (CMAKE_C_FLAGS "-D_POSIX_C_SOURCE=200112L ${CMAKE_C_FLAGS}") 
endif() 
```
That we could leave outside of the std=99 macro.  Would you agree?

File: bld/CMakeLists.txt (hunk 1 above)
Comment (marcschier) [inline bld/CMakeLists.txt]: Minimum required I think is 2.8.11, so if someone has that installed, we need to set the std differently than post ver 3.1, which is when they added the STD variables.  

I am getting curious, 4.8.x works fine (used by travis), I might set up Debian Docker image now and try this.

File: src/pal/pal_net_posix.c [hunk 7]
```diff
@@ -1129,6 +1129,9 @@ int32_t pal_ntop(
     prx_size_t addr_string_size
 )
 {
+#ifndef NI_MAXSERV
+#  define NI_MAXSERV 32 // use same definition as in os_mock.h if not defined
```
Comment (marcschier) [inline src/pal/pal_net_posix.c]: We could also add this to os_posix.h, I wonder though why it is not defined, seems like a standard macro in socket programming...  Maybe I am missing a header.  Did you troll your usr/include to see if it is defined somewhere else or maybe I am missing a __use macro or so... :-(.  Anyway, I will try to repro...

Comment (JuergenKosel) [inline src/pal/pal_net_posix.c]: I have just seen, that /usr/include/netdb.h contains the following:
`# ifdef __USE_MISC`
`#  define NI_MAXHOST      1025`
`#  define NI_MAXSERV      32`
`# endif`
I had created commit 963a5a1cebbf3799bd4c42c9f60b3af06c55a680 before 28ffd768bca96951013dc203fbe3d1958c8d6f33 .
So the modification above may be obsolete or unnecessary.

Comment (marcschier) [inline src/pal/pal_net_posix.c]: Super, then let's just remove.

**Comment by marcschier** [comment]
I have added a Debian dockerfile and can repro this easily.  Still surprised why travis let it pass on trusty.

File: inc/os_mock.h [hunk 8]
```diff
@@ -4,6 +4,12 @@
 #ifndef _os_mock_h_
 #define _os_mock_h_
 
+#include "test_config.h" // Generated by cmake
```
Comment (marcschier) [inline inc/os_mock.h]: -D_HAVE_TIME might have been easier.  I have looked at this , and I think the simplest would be to #define CLOCK_MONOTONIC 4, which it is anyway, and typedef int clockid_t.  That passes for me on debian.

File: bld/test/libtest/CMakeLists.txt [hunk 9]
```diff
@@ -6,6 +6,15 @@
 # not want to pull in submodules for umock, etc.
 #
 
+if(WIN32)
+else()
+    check_include_file("stdint.h" HAVE_STDINT_H)
+    if(HAVE_STDINT_H)
```
Comment (marcschier) [inline bld/test/libtest/CMakeLists.txt]: I would love to learn about under which conditions we need this, i.e. what settings you made on your system, so I can repro this on my end.

File: inc/os_linux.h [hunk 10]
```diff
@@ -9,6 +9,7 @@
 //
 #define _GNU_SOURCE
 #define __USE_GNU
+#define __USE_MISC // needed for SOL_TCP in <netinet/tcp.h>
```
Comment (marcschier) [inline inc/os_linux.h]: I get a warning on strings.h on Ubuntu, strcasecmp not defined, etc.  I will take care of this.

File: .gitignore [hunk 11]
```diff
@@ -17,3 +17,5 @@
 /.vscode/**
 /build/**
 *.user
+# Ignore editor backup files
+*~
```
Comment (marcschier) [inline .gitignore]: Could you submit this on as part of a separate pull request (#7)?

File: bld/test/libtest/CMakeLists.txt (hunk 9 above)
Comment (marcschier) [inline bld/test/libtest/CMakeLists.txt]: Could you submit this on as part of a separate pull request (#7)?

**Comment by marcschier** [comment]
Hi @JuergenKosel, 

I have reproed this on trusty, Jessie, and wheezy.  I have not seen this on the others yet.  I will prepare a more extensive fix that builds on parts of yours, but will include docker build images, script changes to run all distro builds (on Windows), and some other clean up of stuff I encountered (e.g. we should only include "os.h" everywhere).  

If you could cherry-pick the fixes I commented on, either update this pull request or make a new one, then I will pull them in as soon as checks pass.   Apologies for the back and forth and that it takes so long, we very much appreciate your help and interest.

Cheers,

Marc

File: .gitignore (hunk 11 above)
Comment (JuergenKosel) [inline .gitignore]: Okay,
I will split this pull request.
But first I want to have a collection of commits which pass the build tests.

File: inc/os_linux.h (hunk 10 above)
Comment (JuergenKosel) [inline inc/os_linux.h]: I have just read http://man7.org/linux/man-pages/man7/feature_test_macros.7.html
which says that __USE_MISC and __USE_GNU should not be defined here.
It is defined in <features.h> if _BSD_SOURCE or _SVID_SOURCE is defined.
Furthermore a comment in features.h explains:
  

>  The `-ansi' switch to the GNU C compiler, and standards conformance
>    options such as `-std=c99', define __STRICT_ANSI__.  If none of
>    these are defined, or if _DEFAULT_SOURCE is defined, the default is
>    to have _SVID_SOURCE, _BSD_SOURCE, and _POSIX_SOURCE set to one and
>    _POSIX_C_SOURCE set to 200809L.  If more than one of these are
>    defined, they accumulate.  For example __STRICT_ANSI__,
>    _POSIX_SOURCE and _POSIX_C_SOURCE together give you ISO C, 1003.1,
>    and 1003.2, but nothing else.

So I have to modify the above.

Comment (marcschier) [inline inc/os_linux.h]: Yep, same here.  But the biggest problem is that os.h or concrete os_posix.h etc. are included all over the place int the pal, and after some OS headers such as string.h or stdio, etc. are already included, thus feature test macros are already evaluated and these settings do not grab hold.  What I did and what I am testing now is to a) move all os.h to the beginning of the .c, then b) set 
```
//
// Linux OS api used in PAL implementation
//
#define _DEFAULT_SOURCE 1
#define _BSD_SOURCE 1
#define _SVID_SOURCE 1
#define _XOPEN_SOURCE 600
#define _GNU_SOURCE 1
#define _FORTIFY_SOURCE 1
#define _ISOC99_SOURCE 1
#include <features.h>

#include "os_posix.h"
```
I am building this now on all distros and let you know what I find.

**Comment by JuergenKosel** [comment]
This pull request has been obsoleted by #7 
I will create new ones with the remaining changes.
